pyccel/parser/semantic.py

```diff
@@ -3857,31 +3886,37 @@ def _visit_FunctionalFor(self, expr):
         # TODO [YG, 30.10.2020]:
         #  - Check if we should allow the possibility that is_stack_array=True
         # ...
-        lhs_symbol = expr.lhs.base
+        lhs_symbol = expr.lhs
         ne = []
         lhs = self._assign_lhs_variable(lhs_symbol, d_var, rhs=expr, new_expressions=ne)
         lhs_alloc = ne[0]
 
         if isinstance(target, PythonTuple) and not target.is_homogeneous:
-            errors.report(LIST_OF_TUPLES, symbol=expr, severity='error')
+            errors.report(LIST_OF_TUPLES, symbol=expr, severity='fatal')
 
         target.invalidate_node()
 
         loops = [self._visit(i) for i in expr.loops]
-        index = self._visit(index)
+        if expr.index:
+            index = self._visit(index)
 
         l = loops[-1]
+        cnt = 0
         for idx in indices:
             assert isinstance(l, For)
             # Sub in indices as defined here for coherent naming
             if idx.is_temp:
                 self.scope.remove_variable(l.target[0])
                 l.substitute(l.target[0], idx_subs[idx])
-            l = l.body.body[-1]
+            cnt += 1
+            if cnt == len(l.target):
+                l = l.body.body[-1]
+                cnt = 0
```

Comment:
I am surprised that this is not:
```python
            if idx.is_temp:
                self.scope.remove_variable(l.target[cnt])
                l.substitute(l.target[v], idx_subs[idx])
            cnt += 1
            if cnt == len(l.target):
                l = l.body.body[-1]
                cnt = 0
```
Why do you not care about the other targets?

Reply:
Good catch! However, I don't know when this condition applies. I think this is vestigial code back from when `functionalFor` created loop scopes. Could you please come up with an example when idx is temp?

Reply:
`idx` is temp for almost everything except `range` and `enumerate`. The difficulty is finding an example where there are multiple targets. Try this:
```python
import numpy as np

if __name__ == '__main__':
    a = np.array([1,2,3,4])
    b = np.array([5,6,7,8])
    c = [ai*bi for ai, bi in product(a, b)]
```